overlord/devicestate/devicemgr.go

```diff
@@ -528,6 +538,143 @@ func (m *DeviceManager) ensureBootOk() error {
 	return nil
 }
 
+func (m *DeviceManager) ensureCloudInitRestricted() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	if m.cloudInitAlreadyRestricted {
+		return nil
+	}
+
+	var seeded bool
+	err := m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// On Ubuntu Core devices that have been seeded, we want to restrict
+	// cloud-init so that its more dangerous (for an IoT device at least)
+	// features are not exploitable after a device has been seeded. This allows
+	// device administrators and other tools (such as multipass) to still
+	// configure an Ubuntu Core device on first boot, and also allows cloud
+	// vendors to run cloud-init with only a specific data-source on subsequent
+	// boots but disallows arbitrary cloud-init {user,meta,vendor}-data to be
+	// attached to a device via a USB drive and inject code onto the device.
+
+	// TODO: expand the scope to run on any device with a gadget, so i.e.
+	//       classic devices connected to a brand store also have this run?
+	if seeded && !release.OnClassic {
+		opts := &sysconfig.CloudInitRestrictOptions{}
+
+		// check the current state of cloud-init, if it is disabled or already
+		// restricted then we have nothing to do
+		cloudInitStatus, err := cloudInitStatus()
+		if err != nil {
+			return err
+		}
+		statusMsg := ""
+
+		switch cloudInitStatus {
+		case sysconfig.CloudInitDisabledPermanently, sysconfig.CloudInitRestrictedBySnapd:
+			// already been permanently disabled, nothing to do
+			m.cloudInitAlreadyRestricted = true
+			return nil
+		case sysconfig.CloudInitUntriggered:
+			// hasn't been used
+			statusMsg = "reported to be in disabled state"
+		case sysconfig.CloudInitDone:
+			// is done being used
+			statusMsg = "reported to be done"
+		case sysconfig.CloudInitErrored:
+			// cloud-init errored, so we give the device admin / developer a few
+			// minutes to reboot the machine to re-run cloud-init and try again,
+			// otherwise we will disable cloud-init permanently
+			// initialize
```

Comment:
this comment is incomplete, should be:

```suggestion
			// initialize the first time we saw cloud-init in error state
```

(again I will do this in a followup, just raising to myself more than anything